[Merged by Bors] - feat(number_theory): define ramification index and inertia degree

Vierkantor opened the pull request:
We define ramification index `ramification_idx` and inertia degree `inertia_deg` for `P : ideal S` over `p : ideal R` given a ring extension `f : R →+* S`. The literature generally assumes `p` is included in `P`, both are maximal, `R` is the ring of integers of a number field `K` and `S` is the integral closure of `R` in `L`, a finite separable extension of `K`; we relax these assumptions as much as is practical.

---

This PR was split off from #12287.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Vierkantor committed:
feat(number_theory): define ramification index and inertia degree

We define ramification index `ramification_idx` and inertia degree `inertia_deg` for `P : ideal S` over `p : ideal R` given a ring extension `f : R →+* S`. The literature generally assumes `p` is included in `P`, both are maximal, `R` is the ring of integers of a number field `K` and `S` is the integral closure of `R` in `L`, a finite separable extension of `K`; we relax these assumptions as much as is practical.

This PR was split off from #12287.

riccardobrasca commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+  ramification_idx f p P = nat.find h :=
+dif_pos h
+
+lemma ramification_idx_spec (n : ℕ) (hle : map f p ≤ P ^ n) (hgt : ¬ map f p ≤ P ^ (n + 1)) :
```
```suggestion
lemma ramification_idx_spec {n : ℕ} (hle : map f p ≤ P ^ n) (hgt : ¬ map f p ≤ P ^ (n + 1)) :
```

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+
+open_locale big_operators
+open_locale non_zero_divisors
+
```
It looks like you're not using these.
```suggestion
```

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+
+If `P : ideal S` lies over `p : ideal R` for the ring extension `f : R →+* S`
+(assuming `P` and `p` are prime or maximal where needed).
+The **ramification index** `ramification_idx f p P` is the multiplicity of `P` in `map f p`,
```
```suggestion
Given `P : ideal S` lying over `p : ideal R` for the ring extension `f : R →+* S`
(assuming `P` and `p` are prime or maximal where needed),
the **ramification index** `ramification_idx f p P` is the multiplicity of `P` in `map f p`,
```

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+/-
+Copyright (c) 2021 Anne Baanen. All rights reserved.
```
```suggestion
Copyright (c) 2022 Anne Baanen. All rights reserved.
```

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+
+section dec_eq
+
+open_locale classical
```
It would be nice if we could confine `open_locale classical` so that it just contained `ramification_idx` and `inertia_deg` so that the lemmas about these definitions did not explicitly mention `classical.prop_decidable`.

I'd say this is "nice to have" rather than essential. See also relatively recent [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/decidable.20instances.20in.20indicator.20lemmas).

Vierkantor replied:
I agree that in principle we might avoid a conflict by removing the `open_locale` classical. I think in practice the drawbacks outweigh the benefits: Basically all theorems require a huge `decidable` instance if we drop the `classical` locale: `decidable_pred (λ (n : ℕ), ∀ (k : ℕ), map f p ≤ P ^ k → k ≤ n)`. There is also little risk of conflict with an actual decidability instance, since the predicate starts with a `∀` over an infinite set, so you'd have to actually define a specific, nontrivial decision algorithm for this specific problem.

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+If there is no largest such `n` (e.g. because `p = ⊥`), then `ramification_idx` is
+defined to be 0.
+-/
+noncomputable def ramification_idx : ℕ :=
```
I don't think this should be in the root namespace; somebody might well expect this to mean the corresponding concept from complex analysis. I'd guess `ideal` might be the right namespace but you're better-placed than I to decide.

Vierkantor replied:
Good point, I guess `ideal` works well enough.

ocfnash commented on src/number_theory/ramification_inertia.lean:
```diff
@@ -0,0 +1,195 @@
+defined to be 0.
+-/
+noncomputable def ramification_idx : ℕ :=
+if h : ∃ n, map f p ≤ P ^ n ∧ ¬ map f p ≤ P ^ (n + 1)
```
Is it true that:
```lean
lemma ramification_idx_eq_Sup : ramification_idx f p P = Sup { n | map f p ≤ P ^ n } :=
```
Could this be the definition?

Vierkantor replied:
Indeed, it makes for a slightly more powerful definition. Thanks!

Vierkantor committed:
Remove unused locales

Vierkantor committed:
More implicit variables

Vierkantor committed:
Docstring improvements

Vierkantor committed:
Use `nat.has_Sup` to define `ramification_idx`

It costs a bit more work, but in the end everything is equivalent,
and in fact `le_pow_ramification_idx` becomes slightly stronger.

Vierkantor committed:
Move to `ideal` namespace
